### PR TITLE
Add a secret sync directive to the agent

### PR DESCRIPTION
**Pull request by knlambert**
Signed-off-by: Kévin Lambert <kevin.lambert.ca@gmail.com>

## Description

We want to support the sync of a secret from a third party (here ambassador cloud) to automate the creation of Filters.

This PR adds:
* A new field to the directive object, along the one for the rollouts for this use case.
* A handler to add / remove keys from the secret.
* Edit the chart to make it work. Note i've edited an existing role, I think I should rename it, but the CI prevents me from doing it.

I will probably need some hints about how we should manage the version of all of this, especially regarding system-a :)

## Testing

Tested it by deploying it in a kubernetes cluster & interrogating the service to validate it works properly.
Added also a bunch of unit tests to validate the behavior of the command.

## Checklist

 - [x] I made sure to update `CHANGELOG.md`.

   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations

 - [x] This is unlikely to impact how Ambassador performs at scale.

   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability

 - [x] My change is adequately tested.

   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points

 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

 - [x] The changes in this PR have been reviewed for security concerns and adherence to security best practices.

### CODE REVIEW

File: api/agent/director.proto
Comment (LanceEa) [inline api/agent/director.proto:133]: Would we have a specific Result message? I didn't see one...

File: docs/releaseNotes.yml
Comment (LanceEa) [inline docs/releaseNotes.yml:42]: suggestions: 
- use "command" instead of "option" since that is what it is called.
- "synchronize" gives the impression that it is two way. Maybe use "manage" secrets.
- remove "The objective" to more direct language like "This allows"

Maybe something like so:

> Adds a new directive command to the $productName$ Agent service that allows it to manage secrets. This allows a third party product to manage $productName$ CRD's that depend upon a secret.

File: pkg/agent/directive_handler.go
Comment (LanceEa) [inline pkg/agent/directive_handler.go:112]: nit: put space between end of previous func and new func

File: pkg/agent/directive_handler.go
Comment (LanceEa) [inline pkg/agent/directive_handler.go:158]: Rather than introducing a custom interface, would just using the standard `Stringer` interface be sufficient?

Alternatively, a more heavy handed approach would be to update the protocol buffers to have a "BaseCommand" message type that was extended and then use that. But, not sure that is necessary at this time.

File: pkg/agent/directive_handler_test.go
Comment (LanceEa) [inline pkg/agent/directive_handler_test.go:30]: Some additional test coverage would be nice:

1. Set Action with multiple key/value pairs
2. Set Action with no key/value pairs
3. Set command with missing secret, namespace/name
3. Del Action for secret that exists
4. Del Action for secret that does not exist

I would also think some test with Roles missing as well.

File: pkg/agent/directive_handler_test.go
Comment (LanceEa) [inline pkg/agent/directive_handler_test.go:50]: I see this is copying a previous test that already exists but how are we asserting anything to be true? Is it just running this and then manually checking a cluster or ambassador cloud?

File: pkg/agent/secrets.go
Comment (LanceEa) [inline pkg/agent/secrets.go:65]: Since this is using an alias on a string there is nothing stopping someone from creating a `secretSyncCommannd` with action "foo". 

Therefore, we should explicitly catch the case `secretSyncActionSet` and then the default case should probably `panic` because that would mean we have a programming error somewhere and we would want to catch that in CI.

File: pkg/agent/secrets.go
Comment (LanceEa) [inline pkg/agent/secrets.go:129]: typo: `secre` --> `secret`

File: pkg/agent/secrets.go
Comment (LanceEa) [inline pkg/agent/secrets.go:135]: Exported function should have comment

File: pkg/agent/secrets.go
Comment (LanceEa) [inline pkg/agent/secrets.go:144]: I'm guessing we want this to happen because it should never happen unless something was really messed up with the cluster as a whole?

File: pkg/agent/directive_handler_test.go
Comment (LanceEa) [reply pkg/agent/directive_handler_test.go:30]: After looking at `secrets_test.go` some of these are already covered, correct?

File: pkg/agent/rollouts.go
Comment (LanceEa) [inline pkg/agent/rollouts.go:146]: It looks like you moved this to `secrets.go`. Rather than do that should it move to a new file or probably the `utils.go` since it is shared in the package?

**Review by LanceEa — DISMISSED** [review]
In general, it looks inline with what we did with Rollouts. I have left some feedback and comments. 

Definitely want to get a second set of eyes on the permission changes for the roles and want to make sure we have the correct test coverage.

File: api/agent/director.proto
Comment (knlambert) [reply api/agent/director.proto:44]: Oh I'm not. That's just the IDE formating the fields.

File: api/agent/director.proto
Comment (knlambert) [reply api/agent/director.proto:133]: The commands arrive by the `Retrieve` method which is a stream which is always open. 

We don't have a result because it's asynchronous. The commands could take some time to execute, and arrive un bulk. 

Although, when a command is finished, the status is reported with a `CommandResult` object via the method `ReportCommandResult`.

File: charts/emissary-ingress/templates/ambassador-agent.yaml
Comment (knlambert) [reply charts/emissary-ingress/templates/ambassador-agent.yaml:68]: - Yes, the roles are either global or by namespace. When it's by namespace, you can't do anything out of it.
- The agent only creates secrets. The API Keys & Filters are created by the end user using GitOps or manual creation.
- We were planning to only create filters in the ambassador namespace, but I don't know if we should support filters created in a different namespace by the end user. In this case, you're right, this permission may not be enough.

File: pkg/agent/secrets.go
Comment (knlambert) [reply pkg/agent/secrets.go:144]: Yeah, because the library discovers by itself what are the available connection methods in the pod. Usually that's the kubernetes token mounted on a specific path in the pod.

File: pkg/agent/rollouts.go
Comment (knlambert) [reply pkg/agent/rollouts.go:146]: Move to k8s.

File: pkg/agent/directive_handler_test.go
Comment (knlambert) [reply pkg/agent/directive_handler_test.go:30]: 1. Data is passed directly to the function, so doesn't change anything, but I've updated the test case to reflect that.
2. Added test with an empty map, should be allowed though if empty. Not for nil (it will fail in the nil case)
3. It's handled by the directive handler.
4. There is a test case in secrets_test.
5. There is a test case in secrets_test. It doesn't return any error.

File: pkg/agent/directive_handler_test.go
Comment (knlambert) [reply pkg/agent/directive_handler_test.go]: Actually I thought I could actually test something, but the reality is that it is badly designed. I can't replace Agent for an interface or anything which would allow me to test its behavior.

It could for sure be improved, but I don't think that's the purpose of this PR.

File: charts/emissary-ingress/templates/ambassador-agent.yaml
Comment (LanceEa) [reply charts/emissary-ingress/templates/ambassador-agent.yaml:68]: In general, we own the Ambassador namespace so its best if users place stuff in their own namespace. I would think if we were creating the filters on their behalf via the cloud then maybe it makes sense to put them there or some new namespace specific to this use case. 

But in this case if we are creating the `secrets` in our namespace then based on the recommendation of removing the `namespace` from the `APIKeyFilter` will not work if a user creates a Filter in another namespace, correct? Because, we require the secret to be in the same namespace?

File: pkg/agent/directive_handler_test.go
Comment (LanceEa) [reply pkg/agent/directive_handler_test.go]: Ok, that is fine. I don't want to block this but I think that definitely is an area we should return to later to improve our test story here.

**Review by LanceEa — COMMENTED** [review]
For me the big open question is where are these secrets going to get stored and how are they going to be used.

**Comment by knlambert** [comment]
> For me the big open question is where are these secrets going to get stored and how are they going to be used.

The only place these secrets are stored is in this kubernetes secret. We don't keep them in database in the cloud product or anything. We only add or remove key to the secret :)

File: api/agent/director.proto
Comment (LanceEa) [reply api/agent/director.proto:143]: Ok thanks for the detailed update and it makes more sense now. I still find it odd to be a map of always one item but understand what you are doing 😄

File: charts/emissary-ingress/templates/ambassador-agent.yaml
Comment (LanceEa) [reply charts/emissary-ingress/templates/ambassador-agent.yaml:68]: understood

**Comment by LanceEa** [comment]
> > For me the big open question is where are these secrets going to get stored and how are they going to be used.
> 
> The only place these secrets are stored is in this kubernetes secret. We don't keep them in database in the cloud product or anything. We only add or remove key to the secret :)
> 
> Typical workflow is :
> 
> * We give the end user the command to apply the filter (and they does).
> * They creates an API Key in the cloud interface, and keep the API Key for themself.
> * The cloud transmits the key to the agent, which adds it as an entry in the secret used by the filter (and creates the secret if it if it doesn't exist).
> * When the end user revokes the API Key, we remove the entry from the secret.
> 
> That's it.

Ok the part that is tripping me up, that I guess I want to clarify is that in this other PR https://github.com/emissary-ingress/emissary/pull/4312/files#diff-ff02720b66a6ee517bcd28beddc007b390071e1ca4a83d3d6bacc5e0883cc3e3R398 we are saying we will only load secrets if the Filter is in the same namespace as the Filter. 

So, does this mean the agent needs access to create secrets in all namespaces or are we just saying to the customer we will create the secret with APIKeys within the `ambassador` namespace and you must setup your gitops to add the Filter to that namespace too?

**Comment by LanceEa** [comment]
BTW...you also have a merge conflict now due to a recent merge into master

**Comment by knlambert** [comment]
> So, does this mean the agent needs access to create secrets in all namespaces or are we just saying to the customer we will create the secret with APIKeys within the ambassador namespace and you must setup your gitops to add the Filter to that namespace too?

It only needs access to create secrets in the ambassador namespace (except if you think we should anticipate that it may be in others), and the clients will configure their GitOps to create the filters in that ambassador namespace too.

**Comment by LanceEa** [comment]
> It only needs access to create secrets in the ambassador namespace (except if you think we should anticipate that it may be in others), and the clients will configure their GitOps to create the filters in that ambassador namespace too (since the secrets are there).

That is fine for a v0 implementation but I do believe that users will want more flexibility in the future. The purpose of most of the Emissary-ingress CRD's is to empower the end developer and provide them with flexibility without having to involve a platform engineer. I could see where platform engineers are more strict about letting developers add things to the `ambassador` namespace while requiring them to create Mappings, Filters, Hosts, etc... in their developer specific namespace. 

However, to enable this you would have to give the Agent access to manage secrets in all namespaces and the cloud interface would have to add the secret to the correct namespace that the filter was added to so this could be a source of friction due to Filters being added to different namespaces.

So, just make sure the v0 implementation is well documented so users understand the current restrictions.

File: pkg/agent/secrets.go
Comment (LanceEa) [inline pkg/agent/secrets.go:81]: Do we really want the agent crashing when this happens?

File: pkg/agent/secrets.go
Comment (LanceEa) [inline pkg/agent/secrets.go:22]: Why is this being exported now? 

If it needs to be exported then it should have doc comments

**Review by LanceEa — COMMENTED** [review]
a few last questions on the recent changes and then I want a second review on this. I will ping someone on the team to take a second look.

File: pkg/agent/secrets.go
Comment (knlambert) [reply pkg/agent/secrets.go:22]: Because otherwise `func NewSecretsGetter(namespace string) (secretInterface, error)` returns a private interface, and it triggers a check from the IDE saying `Exported function with the unexported return type`

**Comment by LanceEa** [comment]
:( changelog merge conflicts due to merging your other PR. Fix that up and once we a second review then I will be good on this one.

**Comment by LanceEa** [comment]
Looks like maybe you need to re-run `make generate` after a rebase?

**Review by LanceEa — APPROVED** [review]
lgtm...I want to get a second set of eye's on the RBAC changes and then we should be good.

**Review by alexgervais — DISMISSED** [review]
I agree with @LanceEa, this is fine for a v0 implementation but users will want more flexibility in the future. Given Edge Stack already has a ClusterRole to read and write secret values (acme tls), I don't foresee pushback from adding the same permission to the agent component.

LGTM, if:
- The secret namespace limitation is documented in the public docs.
- We capture a story future consideration to remove the limitation.

**Comment by knlambert** [comment]
> I agree with @LanceEa, this is fine for a v0 implementation but users will want more flexibility in the future. Given Edge Stack already has a ClusterRole to read and write secret values (acme tls)

Then, would it be better to just add the agent to this role ? Then we'll be able to update the code to create filters in other namespaces if it's needed without having to release again.

**Comment by alexgervais** [comment]
@knlambert I would keep both roles distinct. It's nice to have the agent role be different than Emissary (especially with what's coming up next cycle). We'd have to add new rules to the existing agent role.

**Comment by LanceEa** [comment]
Does it make more sense to break new commits up into new PR's? IMO, we had already approved this so personally I would have preferred the latest commit to be in a new PR.

File: pkg/agent/directive_handler.go
Comment (LanceEa) [inline pkg/agent/directive_handler.go]: Why is this ok now? I would have thought the secret would need to be required when executing the Set Command?

Comment (knlambert) [reply pkg/agent/directive_handler.go]: Hello Lance, sorry about that, we should be ok now, we identified a last use case where we need the secret to be empty, and even if we give an empty map which is not `nil`, GRPC converts that into a `nil value`. That's why I've removed this condition.

File: pkg/agent/secrets.go
Comment (knlambert) [inline pkg/agent/secrets.go:62]: I've changed this for interface{} because it needs to be an object if the value is {}. "{}" won't work.

**Review by LanceEa — APPROVED** [review]
lgtm

**Review by LanceEa — APPROVED** [review]

**Pull request merged**
